### PR TITLE
Dashboard: Fix Table view when editing causes the panel data to not update

### LINKED ISSUE #34820: Table view when editing causes the panel data to not update
**What happened**:

Setting visualisation to "Table view" when editing causes the panel data to not update.

**What you expected to happen**:

Updating a query when in the new "Table view" mode should still show new data as it's updated. 

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a new panel using gdev-testdata data source, configured as:
   - Scenario: CSV Metric Values
   - String Input: hello_world
2. Configure the panel to use the Table visualisation
3. Change the 'String Input' in query editor
   - _Observe that the value updates correctly in the visualisation_ 😄 
4. Turn on the "Table view" mode
5. Change the 'String Input' in query editor
   - _Observe that the value does not update in the "table view" visualisation_ 😢 
6. Turn "Table view" mode back off
   - _Observe that the value still has not updated in the visualisation_ 😐
3. Change the 'String Input' in query editor
   - _Observe that the value updates correctly in the visualisation_ 😄 
  
I also replicated this issue using similar steps with the Azure Monitor data source, so the issue seems to be independent to the data source.

_Originally posted by @joshhunt in https://github.com/grafana/grafana/issues/34624#issuecomment-847703102_

### PR DESCRIPTION
**Pull request by axelavargas**
**What this PR does / why we need it**:

Setting visualization to "Table view" when editing causes the panel data to not update. This is because the `PanelEditorTableView` was not subscribing to the `RefreshEvents`

**Which issue(s) this PR fixes**:

Fixes #34820

**Special notes for your reviewer**:

The approach I took was extracting from the `PanelChrome` component the logic of running all the panel queries and adding the subscription of the  `RefreshEvent` in the `PanelEditorTableView`

TableViewWithUpdates

### CODE REVIEW

File: public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx [hunk 1]
```diff
@@ -1,23 +1,43 @@
...
     frameIndex: 0,
     showHeader: true,
   });
 
+  const timeSrv: TimeSrv = getTimeSrv();
+  const timeData = applyPanelTimeOverrides(panel, timeSrv.timeRange());
+
+  const onRefresh = useCallback(() => {
+    panel.runAllPanelQueries(dashboard.id, dashboard.getTimezone(), timeData, width);
+  }, [dashboard, panel, timeData, width]);
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx:30]: Maybe the timeSrv lines should be moved inside the `useCallback`
```suggestion
  const onRefresh = useCallback(() => {
    const timeSrv: TimeSrv = getTimeSrv();
    const timeData = applyPanelTimeOverrides(panel, timeSrv.timeRange());
    panel.runAllPanelQueries(dashboard.id, dashboard.getTimezone(), timeData, width);
  }, [dashboard, panel, width]);
```

File: public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx [hunk 2]
```diff
@@ -1,23 +1,43 @@
...
+
+  // Subscribe to panel events
+  useEffect(() => {
+    const subs = new Subscription();
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx:34]: Not sure how `subs` will be treated when this is out of scope? Do you know? Maybe we should create a proper ref instead? See https://github.com/grafana/grafana/blob/fcdfa880899a0b7080234166da97a2eec0f5868d/public/app/features/dashboard/components/PanelEditor/usePanelLatestData.ts#L22

File: public/app/features/dashboard/state/PanelModel.ts [hunk 3]
```diff
@@ -292,6 +293,23 @@ export class PanelModel implements DataConfigSource {
...
+      timeRange: timeData.timeRange,
+      timeInfo: timeData.timeInfo,
+      maxDataPoints: this.maxDataPoints || width,
+      minInterval: this.interval,
+      scopedVars: this.scopedVars,
+      cacheTimeout: this.cacheTimeout,
+      transformations: this.transformations,
+    });
+  }
+
```
Comment (hugohaggmark) [inline public/app/features/dashboard/state/PanelModel.ts:312]: Hmm, I'm not 100% sure this is a good idea. Now we'll have both `getQueryRunner` and `runAllPanelQueries` on PanelModel that can be used to do the same.

**Review by hugohaggmark — COMMENTED** [review]
Skilled work on this and works as expected 🎉  I left some nits for consideration.

File: public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx (hunk 2 above)
Comment (torkelo) [reply public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx:34]: why? this is not out of scope anywhere? this is how we do it everywhere else we create subscriptions inside a useEffect. Never used a ref for it, not sure why we would need to, it's not used outside the useEffect. 

But we not need to use Subscription here as we only have one Unsubscribable, so we can just do: 

```ts
useEffect(() => {

    const sub = panel.events.subscribe(RefreshEvent, () => {
        panel.runAllPanelQueries(dashboard.id, dashboard.getTimezone(), timeData, width);
    }));

    return () => sub.unsubscribe;
}, [onRefresh, panel, dashboard, timeData, width]);
```

**Review by torkelo — COMMENTED** [review]
Awesome job! looks good. 

Only nitpick is the onRefresh with useCallback, we can simplify this with doing everything in useEffect:

```ts
useEffect(() => {

    const sub = panel.events.subscribe(RefreshEvent, () => {
        panel.runAllPanelQueries(dashboard.id, dashboard.getTimezone(), timeData, width);
    }));

    return () => sub.unsubscribe;
}, [onRefresh, panel, dashboard, timeData, width]);
```

Spitting it up into one useCallback and one useEffect looks unessary and since the useEffect depends on the onRefresh callback it will be invaldiated every time onRefresh is invalidated so there is nothing gained by splitting them

File: public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx (hunk 2 above)
Comment (hugohaggmark) [reply public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx:34]: You're right Torkel, I was reading it wrong.

**Comment by hugohaggmark** [comment]
> Awesome job! looks good.
> 
> Only nitpick is the onRefresh with useCallback, we can simplify this with doing everything in useEffect:
> 
> ```ts
> useEffect(() => {
> 
>     const sub = panel.events.subscribe(RefreshEvent, () => {
>         panel.runAllPanelQueries(dashboard.id, dashboard.getTimezone(), timeData, width);
>     }));
> 
>     return () => sub.unsubscribe;
> }, [onRefresh, panel, dashboard, timeData, width]);
> ```
> 
> Spitting it up into one useCallback and one useEffect looks unessary and since the useEffect depends on the onRefresh callback it will be invaldiated every time onRefresh is invalidated so there is nothing gained by splitting them

I agree that this is a better solution 👍

File: public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx (hunk 1 above)
Comment (dprokop) [reply public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx:30]: yup, this makes sense ^^

File: public/app/features/dashboard/state/PanelModel.ts (hunk 3 above)
Comment (dprokop) [reply public/app/features/dashboard/state/PanelModel.ts:312]: I agree with Hugo about this, think we don't need this abstraction.

**Review by dprokop — APPROVED** [review]
@axelavargas great work on this one, it wasn't trivial! Agree with Hugo's comments, whene these are fixed we should be ready to merge :)

File: public/app/features/dashboard/state/PanelModel.ts (hunk 3 above)
Comment (axelavargas) [reply public/app/features/dashboard/state/PanelModel.ts:312]: 🤔 you have a good point, but I am not sure, the idea of creating this abstraction was to be able to reuse it on `PanelChrome` and `PanelEditorTableView` components without passing all the panel properties that the `run` function needs.

If I remove the abstraction, it means we will use directly the `getQueryRunner().run`  in those components and we need to pass all the panel props, is it the approach we want to follow? @hugohaggmark @dprokop @torkelo

File: public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx (hunk 1 above)
Comment (axelavargas) [reply public/app/features/dashboard/components/PanelEditor/PanelEditorTableView.tsx]: thanks for the suggestion 🙌🏾  https://github.com/grafana/grafana/pull/34998/commits/583d7c4fea42b43c8b4a36202a572ecbec13a234#diff-0b63135356fcc7e14a567cfe8ac6f138a92b792c8622aa2a1aa9e6d845e88337R27 🚀

**Comment by axelavargas** [comment]
> > Awesome job! looks good.
> > Only nitpick is the onRefresh with useCallback, we can simplify this with doing everything in useEffect:
> > ```ts
> > useEffect(() => {
> > 
> >     const sub = panel.events.subscribe(RefreshEvent, () => {
> >         panel.runAllPanelQueries(dashboard.id, dashboard.getTimezone(), timeData, width);
> >     }));
> > 
> >     return () => sub.unsubscribe;
> > }, [onRefresh, panel, dashboard, timeData, width]);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Spitting it up into one useCallback and one useEffect looks unessary and since the useEffect depends on the onRefresh callback it will be invaldiated every time onRefresh is invalidated so there is nothing gained by splitting them
> 
> I agree that this is a better solution

Thanks for the feedback! indeed this is a simpler approach :D https://github.com/grafana/grafana/pull/34998/commits/583d7c4fea42b43c8b4a36202a572ecbec13a234 🚀

File: public/app/features/dashboard/state/PanelModel.ts (hunk 3 above)
Comment (hugohaggmark) [reply public/app/features/dashboard/state/PanelModel.ts:312]: Aha, I see what you mean know @axelavargas but aren't the properties needed already on `panel` & `dashboard`? But I'm not on the fences on this one so you can leave it like this.

**Review by hugohaggmark — APPROVED** [review]
Outstanding work on this and works as expected! 🎉

**Pull request merged**